### ISSUE TITLE
Different results using same code but different CPU type

### ISSUE THREAD

**Issue by rluethy** [keras-team/keras#15740]
Reposted from TF issues: https://github.com/tensorflow/tensorflow/issues/53277

**System information**

Linux Ubuntu 20.04
TensorFlow installed from (source or binary): binary
TensorFlow version (use command below): 2.5.0
Python version: 3.8

**Describe the current behavior**

When tuning TF models we noted that we were getting different results when using different instance types on AWS. We started removing possible sources of variability until the only difference was the CPU type of the instances. We created a minimal example that trains a simple MLP on a training set and predicts probabilities for a validation set. The following figure compares the predictions obtained with an Intel CPU to those with an AMD CPU:

![image](https://user-images.githubusercontent.com/3641872/144479986-da090d04-6237-44c2-85e9-7e1684838505.png)

**Describe the expected behavior**

The predicted probabilities should be the same with both CPU types and the points in the above plot should all be on the diagonal.

**Standalone code to reproduce the issue**

Code to reproduce issue: https://github.com/rluethy/tf-cpu-type-diff

**Comment by old-school-kid** [comment]
Can you please a take a look at [this](https://stackoverflow.com/questions/43221730/tensorflow-same-code-but-get-different-result-from-cpu-device-to-gpu-device)

**Comment by rluethy** [comment]
Thanks for the link. That issue is similar, but it's the difference between CPU and GPU which have different precision. The difference here is between Intel and AMD CPUs, which both have the same precision. 
Also in the example code all seeds are initialized with a known value, random states should be identical.

**Comment by sachinprasadhs** [comment]
Could you please run this comparison in Tensorflow 2.7 version and let us know if you still find any difference. Thanks!

**Comment by rluethy** [comment]
The differences with Tensorflow are present, but much smaller. Top differences are:

106   0.000012
171   0.000006
131   0.000005
27    0.000004
111   0.000004
178   0.000003
15    0.000003
149   0.000003
141   0.000003
119   0.000002
Name: prob_1, dtype: float64

The difference are not visible on the graph:

![6c43e304-9749-4924-bd95-add2378ee6f2](https://user-images.githubusercontent.com/3641872/145488238-07b3b70d-ced7-48ee-991c-07eabd292d4b.png)

**Comment by old-school-kid** [comment]
The code provided uses a seed for initialization in layers. [This issue](https://github.com/keras-team/keras/issues/15586#issue-1045027853) points out that an initialization when called multiple times produces different result. Workaround for now can be set the deterministic true which is experimental right now.
@jvishnuvardhan can you please take a look at this issue and the one I have mentioned. 
Thanks!

**Comment by ljbuturovic** [comment]
Hi @old-school-kid : 1. [the issue](https://github.com/keras-team/keras/issues/15586#issue-1045027853) refers to problem with multiple initializers, as you point out. But the code in this ticket uses single initializer. 2. [the issue](https://github.com/keras-team/keras/issues/15586#issue-1045027853) applies to TF2.7. But this ticket refers to TF2.5.

**Comment by rchao** [comment]
Note: This appears to have been fixed in 2.7 and we can't retroactively patch the fix into older releases (2.5 here).
